File: bfs_dfs/pick_numbers.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int numbers[101];
bool isAns(int init, int node, bool visit[], int *cnt) {
	//cout << node << visit[node] << endl;
	if (init == node and visit[node]) {
		return true;
	}
	if (visit[node]) return false;

	visit[node] = true;
	*cnt += 1;
	return isAns(init, numbers[node], visit, cnt);
}
int main(void) {
	cin.tie(0);
	ios::sync_with_stdio;

	vector<int> ans;
	int total = 0;
	int n;
	cin >> n;
	
	bool visit[101] = { false };

	for (int i = 1; i <= n; i++) {
		cin >> numbers[i];
	}

	for (int i = 1; i <= n; i++) {
		if (visit[i]) continue;
		bool nvisit[101] = {false};
		int cnt = 0;

		for (int j = 1; j <= n; j++) {
			if (!visit[j]) continue;
			nvisit[j] = true;
		}

		if (isAns(i, i, nvisit, &cnt)) {
			for (int j = 1; j <= n; j++) {
				if (visit[j] == nvisit[j]) continue;
				visit[j] = true;
				ans.push_back(j);
			}
			total += cnt;
		}
	}
	sort(ans.begin(), ans.end());
	cout << total << endl;
	for (int i = 0; i < total; i++) {
		cout << ans[i] << endl;
	}
	return 0;
}

File: bfs_dfs/pick_numbers.py
import sys

input = sys.stdin.readline

N = int(input().rstrip())

numbers = [int(input().rstrip()) for _ in range(N)]
numbers.insert(0,0)
visit = [False] * (N + 1)

def check(init, node, visit, cnt) :
    if visit[node] and init == node : return True, cnt
    if visit[node] : return False, cnt
    visit[node] = True
    nextNode = numbers[node]

    return check(init, nextNode, visit, cnt + 1)

ans = 0
ansNumbers = []
for i in numbers :
    if visit[i] or i == 0 : continue
    nvisit = visit.copy()
    result, cnt = check(i, i, nvisit, 0)
    if result :
        for i in range(1, N+1):
            if visit[i] != nvisit[i] :
                ansNumbers.append(i)
                visit[i] = True
        ans += cnt

print(ans)
ansNumbers.sort()
for i in ansNumbers :
    print(i)